File: 84. Largest Rectangle in Histogram.java

import java.util.Stack;

class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack();
        int n = heights.length;
        int left[] = new int[n];
        int right[] = new int[n];
        for (int i = 0; i < n; ++i) left[i] = -1;
        for (int i = 0; i < n; ++i) right[i] = n;
        int res = 0;
        for(int i = 0; i < n; ++i){
            while(!stack.isEmpty() && heights[stack.peek()] >= heights[i]){
                right[stack.peek()] = i;
                stack.pop();
            }
            if(!stack.isEmpty()) left[i] = stack.peek();
            stack.push(i);
        }
        for(int i = 0; i < n; ++i){
            res = Math.max(res, heights[i] * (right[i] - left[i] - 1));
        }
        return res;
    } 
}

File: 84. Largest Rectangle in Histogram.py

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        left = [-1] * len(heights)
        right = [len(heights)] * len(heights)
        res = 0
        for i, num in enumerate(heights):
            while stack and heights[stack[-1]] >= num:
                right[stack[-1]] = i
                stack.pop()
            if stack:
                left[i] = stack[-1]
            stack.append(i)

        for i in range(len(heights)):
            res = max(res, heights[i] * (right[i] - left[i] - 1))
        
        return res